src/tracks/playabletrack/wavetrack/ui/ChangeClipSpeedDialog.cpp

```diff
@@ -11,106 +11,237 @@
+class ScopedInvisiblePanel : public ScopedSizer<wxPanel*, int>
+{
+public:
+   ScopedInvisiblePanel(ShuttleGui& s, int border = 0)
+       : ScopedSizer<wxPanel*, int>(
+            s, &ShuttleGui::StartInvisiblePanel, &ShuttleGui::EndInvisiblePanel,
+            std::move(border))
+   {
+   }
+};
+
+class ScopedVerticalLay : public ScopedSizer<void, int>
+{
+public:
+   ScopedVerticalLay(ShuttleGui& s, int iProp = 1)
+       : ScopedSizer<void, int>(
+            s, &ShuttleGui::StartVerticalLay, &ShuttleGui::EndVerticalLay,
+            std::move(iProp))
+   {
+   }
+};
+} // namespace
+
 ChangeClipSpeedDialog::ChangeClipSpeedDialog(
-   WaveTrack& waveTrack, WaveTrack::Interval& interval, wxWindow* parent)
+   bool playbackOngoing, WaveTrack& waveTrack, WaveTrack::Interval& interval,
+   wxWindow* parent)
     : wxDialogWrapper(
-         parent, wxID_ANY, XO("Change Speed"), wxDefaultPosition, { 480, 250 },
-         wxDEFAULT_DIALOG_STYLE)
+         parent, wxID_ANY, XO("Pitch And Speed"), wxDefaultPosition,
+         { 480, 250 }, wxDEFAULT_DIALOG_STYLE)
+    , mPlaybackOngoing { playbackOngoing }
     , mTrack { waveTrack }
     , mTrackInterval { interval }
     , mClipSpeed { 100.0 / interval.GetStretchRatio() }
     , mOldClipSpeed { mClipSpeed }
 {
+   const auto shift = mTrackInterval.GetSemitoneShift();
+   mShift.semis = std::floor(shift);
+   mShift.cents = std::round((shift - mShift.semis) * 100);
+   mOldShift = mShift;
+
    ShuttleGui s(this, eIsCreating);
 
-   s.StartVerticalLay();
    {
+      ScopedVerticalLay v { s };
       PopulateOrExchange(s);
    }
-   s.EndVerticalLay();
 
    // TODO: Tolerance?
-   // Stretch ratio of 
+   // Stretch ratio of
    assert(mOldClipSpeed > 0.0);
 
    Layout();
    Fit();
    Centre();
 
-   Bind(
-      wxEVT_CHAR_HOOK,
-      [this](auto& evt)
+   Bind(wxEVT_CHAR_HOOK, [this](auto& evt) {
+      if (!IsEscapeKey(evt))
       {
-         if (!IsEscapeKey(evt))
-         {
-            evt.Skip();
-            return;
-         }
+         evt.Skip();
+         return;
+      }
 
-         EndModal(wxID_CANCEL);
-      });
+      OnCancel();
+   });
 }
 
 ChangeClipSpeedDialog::~ChangeClipSpeedDialog() = default;
 
 void ChangeClipSpeedDialog::PopulateOrExchange(ShuttleGui& s)
 {
-   s.StartInvisiblePanel(15);
-   s.StartVerticalLay(wxEXPAND, 1);
    {
-      s.SetBorder(1);
-      s.StartHorizontalLay(wxEXPAND);
+      ScopedInvisiblePanel panel { s, 15 };
+      s.SetBorder(0);
       {
-         s.AddPrompt(XO("Clip speed"));
-         s.AddSpace(0, 0, 1);
-         s.Name(XO("Clip speed"))
-            .NameSuffix(Verbatim("%"))
-            .TieNumericTextBox({}, mClipSpeed, 14, true)
-            ->Bind(wxEVT_TEXT_ENTER, [this](auto&) { OnOk(); });
+         ScopedHorizontalLay h { s, wxLeft };
+         s.AddPrompt(XO("Clip Pitch"));
+      }
+      {
+         ScopedHorizontalLay h { s, wxLeft };
+         constexpr auto minSemis = -24;
+         constexpr auto maxSemis = 60;
```

Comment:
Two octaves down, five octaves up -- might comment somewhere that these are the bounds.  Maybe more prominently in the header file.

Reply:
Adding a comment.